src/box/memtx_engine.cc

```diff
@@ -333,7 +363,12 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	RegionGuard region_guard(&fiber()->gc);
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
-	*is_space_system = (request.space_id < BOX_SYSTEM_ID_MAX);
+	if (!space_id_is_system(request.space_id)) {
+		if (complete_system_space_recovery(state) != 0)
+			return -1;
+	} else {
+		*state = RECOVERING_SYSTEM_SPACES;
```

Comment:
The code'd look nicer if the state was updated in one place, because one could see the state-diagram and expectations by looking at just one small function:

<details><summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 2fdaf29fd3f0..ce8f687ce633 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -231,6 +231,29 @@ enum snapshot_recovery_state {
 	DONE_RECOVERING_SYSTEM_SPACES,
 };
 
+/** Checks and updates the snapshot recovery state. */
+static int
+snapshot_recovery_state_update(enum snapshot_recovery_state *state,
+			       bool is_system_space_row)
+{
+	switch (*state) {
+	case SNAPSHOT_RECOVERY_NOT_STARTED:
+		if (!is_system_space_row) {
+			diag_set(ClientError, ER_MISSING_SYSTEM_SPACES);
+			return -1;
+		}
+		*state = RECOVERING_SYSTEM_SPACES;
+		break;
+	case RECOVERING_SYSTEM_SPACES:
+		if (!is_system_space_row)
+			*state = DONE_RECOVERING_SYSTEM_SPACES;
+		break;
+	case DONE_RECOVERING_SYSTEM_SPACES:
+		break;
+	}
+	return 0;
+}
+
 /**
  * Recovers one xrow from snapshot.
  *
@@ -328,20 +351,6 @@ memtx_engine_recover_synchro(const struct xrow_header *row)
 	return txn_limbo_process(&txn_limbo, &req);
 }
 
-/**
- * The snapshot must contain system spaces.
- */
-static inline int
-complete_system_space_recovery(enum snapshot_recovery_state *state)
-{
-	if (*state == SNAPSHOT_RECOVERY_NOT_STARTED) {
-		diag_set(ClientError, ER_MISSING_SYSTEM_SPACES);
-		return -1;
-	}
-	*state = DONE_RECOVERING_SYSTEM_SPACES;
-	return 0;
-}
-
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 				  struct xrow_header *row,
@@ -349,7 +358,8 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 {
 	assert(row->bodycnt == 1); /* always 1 for read */
 	if (row->type != IPROTO_INSERT) {
-		if (complete_system_space_recovery(state) != 0)
+		if (snapshot_recovery_state_update(
+				state, /*is_system_space_row=*/false) != 0)
 			return -1;
 		if (row->type == IPROTO_RAFT)
 			return memtx_engine_recover_raft(row);
@@ -363,12 +373,9 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	RegionGuard region_guard(&fiber()->gc);
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
-	if (!space_id_is_system(request.space_id)) {
-		if (complete_system_space_recovery(state) != 0)
-			return -1;
-	} else {
-		*state = RECOVERING_SYSTEM_SPACES;
-	}
+	bool is_system_space_row = space_id_is_system(request.space_id);
+	if (snapshot_recovery_state_update(state, is_system_space_row) != 0)
+		return -1;
 	struct space *space = space_cache_find(request.space_id);
 	if (space == NULL)
 		goto log_request;
@@ -666,10 +673,10 @@ memtx_engine_bootstrap(struct engine *engine)
 		return -1;
 
 	int rc;
-	enum snapshot_recovery_state unused;
 	struct xrow_header row;
+	enum snapshot_recovery_state state = SNAPSHOT_RECOVERY_NOT_STARTED;
 	while ((rc = xlog_cursor_next(&cursor, &row, true)) == 0) {
-		rc = memtx_engine_recover_snapshot_row(memtx, &row, &unused);
+		rc = memtx_engine_recover_snapshot_row(memtx, &row, &state);
 		if (rc < 0)
 			break;
 	}
```

</details>

Reply:
Thanks, applied your suggestion!